### PR TITLE
feat: evalutorBase adds networktopology

### PR DESCRIPTION
**Pull request by MinH-09**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

**Review by gaius-qi — CHANGES_REQUESTED** [review]
Add new file `evaluator_network_topology.go`

File: scheduler/scheduling/evaluator/evaluator_network_topology.go [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+	logger "d7y.io/dragonfly/v2/internal/dflog"
+	"d7y.io/dragonfly/v2/pkg/math"
+	"d7y.io/dragonfly/v2/pkg/types"
+	"d7y.io/dragonfly/v2/scheduler/networktopology"
+	"d7y.io/dragonfly/v2/scheduler/resource"
+)
+
+const (
+	// Finished piece weight.
+	networkTopologyPartFinishedPieceWeight float64 = 0.2
```
Comment (gaius-qi) [inline scheduler/scheduling/evaluator/evaluator_network_topology.go:36]: networkTopologyPart naming is wrong

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: scheduler/scheduling/evaluator/evaluator_network_topology.go (hunk 1 above)
Comment (MinH-09) [reply scheduler/scheduling/evaluator/evaluator_network_topology.go]: Now I change it to networkTopologyBasedAlgorithmFinishedPieceWeight. How about it?

**Review by gaius-qi — CHANGES_REQUESTED** [review]

**Review by gaius-qi — CHANGES_REQUESTED** [review]
1. Refactor options in scheduling.
2. Refactor common `IsBadNode`.

File: scheduler/config/config.go [hunk 2]
```diff
@@ -639,23 +635,23 @@ func (cfg *Config) Validate() error {
 		}
 	}
 
-	if cfg.NetworkTopology.CollectInterval <= 0 {
+	if cfg.Scheduler.NetworkTopology.CollectInterval <= 0 {
```
Comment (gaius-qi) [inline scheduler/config/config.go:638]: Condition cfg.Scheduler.Algorithm equals evaluator.NetworkTopologyAlgorithm must be met to check `NetworkTopology` config.

File: scheduler/scheduling/evaluator/evaluator.go [hunk 3]
```diff
@@ -17,20 +17,40 @@
...
+	// If the number of samples is greater than or equal to 30,
+	// it is close to the normal distribution.
+	normalDistributionLen = 30
+
+	// When costs len is greater than or equal to 2,
+	// the last cost can be compared and calculated.
+	minAvailableCostLen = 2
+)
+
+type BaseEvaluator struct{}
```
Comment (gaius-qi) [inline scheduler/scheduling/evaluator/evaluator.go:52]: Rename `BaseEvaluator` to `Evaluator`.

File: scheduler/scheduling/evaluator/evaluator_network_topology.go [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+	// IDC affinity weight.
+	networkTopologyIDCAffinityWeight = 0.11
+
+	// Location affinity weight.
+	networkTopologyLocationAffinityWeight = 0.11
+)
+
+const (
+	// Maximum score.
+	networkTopologyMaxScore float64 = 1
```
Comment (gaius-qi) [inline scheduler/scheduling/evaluator/evaluator_network_topology.go:55]: Move `maxScore` and `minScore` from `evaluator_base.go` to `evaluator.go` and use common `maxScore` and `minScore`.

File: scheduler/scheduling/evaluator/evaluator_network_topology.go [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+	networkTopologyFinishedPieceWeight float64 = 0.2
+
+	// Parent's host upload success weight.
+	networkTopologyParentHostUploadSuccessWeight = 0.2
+
+	// Free upload weight.
+	networkTopologyFreeUploadWeight = 0.15
+
+	// Network topology weight.
+	networkTopologyNetworkTopologyWeight = 0.12
```
Comment (gaius-qi) [inline scheduler/scheduling/evaluator/evaluator_network_topology.go:41]: `networkTopologyProbeWeight`

File: scheduler/scheduling/evaluator/evaluator_network_topology.go [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+	// defaultPingTimeout specifies a default timeout before ping exits.
+	defaultPingTimeout = 1 * time.Second
+)
+
+type evaluatorNetworkTopology struct {
+	BaseEvaluator
+	networktopology networktopology.NetworkTopology
+}
+
+type Option func(en *evaluatorNetworkTopology)
```
Comment (gaius-qi) [inline scheduler/scheduling/evaluator/evaluator_network_topology.go:74]: `type Option func(e *evaluatorNetworkTopology)`

File: scheduler/scheduling/evaluator/evaluator_network_topology.go [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+
+// WithNetworkTopology sets the networkTopology.
+func WithNetworkTopology(networktopology networktopology.NetworkTopology) Option {
+	return func(en *evaluatorNetworkTopology) {
+		en.networktopology = networktopology
+	}
+}
+
+func NewEvaluatorNetworkTopology(options ...Option) Evaluator {
+	en := &evaluatorNetworkTopology{}
```
Comment (gaius-qi) [inline scheduler/scheduling/evaluator/evaluator_network_topology.go:84]: `e := &evaluatorNetworkTopology{}`

File: scheduler/scheduling/evaluator/evaluator_network_topology.go [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+type evaluatorNetworkTopology struct {
+	BaseEvaluator
+	networktopology networktopology.NetworkTopology
+}
+
+type Option func(en *evaluatorNetworkTopology)
+
+// WithNetworkTopology sets the networkTopology.
+func WithNetworkTopology(networktopology networktopology.NetworkTopology) Option {
+	return func(en *evaluatorNetworkTopology) {
```
Comment (gaius-qi) [inline scheduler/scheduling/evaluator/evaluator_network_topology.go:78]: `e *evaluatorNetworkTopology`

File: scheduler/scheduling/evaluator/evaluator_network_topology.go [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+		networkTopologyParentHostUploadSuccessWeight*networkTopologyCalculateParentHostUploadSuccessScore(parent) +
+		networkTopologyFreeUploadWeight*networkTopologyCalculateFreeUploadScore(parent.Host) +
+		networkTopologyHostTypeWeight*networkTopologyCalculateHostTypeScore(parent) +
+		networkTopologyIDCAffinityWeight*networkTopologyCalculateIDCAffinityScore(parentIDC, childIDC) +
+		networkTopologyLocationAffinityWeight*networkTopologyCalculateMultiElementAffinityScore(parentLocation, childLocation) +
+		networkTopologyNetworkTopologyWeight*en.networkTopologyCalculateNetworkTopologyScore(parent.ID, child.ID)
+}
+
+// networkTopologyCalculatePieceScore 0.0~unlimited larger and better.
+func networkTopologyCalculatePieceScore(parent *resource.Peer, child *resource.Peer, totalPieceCount int32) float64 {
```
Comment (gaius-qi) [inline scheduler/scheduling/evaluator/evaluator_network_topology.go:121]: `func (e *evaluatorNetworkTopology) calculatePieceScore(dst, src string) float64 {`

File: scheduler/scheduling/evaluator/evaluator_network_topology.go [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+
+	// Use the difference between the parent node and the child node to
+	// download the piece to roughly represent the piece score.
+	parentFinishedPieceCount := parent.FinishedPieces.Count()
+	childFinishedPieceCount := child.FinishedPieces.Count()
+	return float64(parentFinishedPieceCount) - float64(childFinishedPieceCount)
+}
+
+// networkTopologyCalculateParentHostUploadSuccessScore 0.0~unlimited larger and better.
+func networkTopologyCalculateParentHostUploadSuccessScore(peer *resource.Peer) float64 {
```
Comment (gaius-qi) [inline scheduler/scheduling/evaluator/evaluator_network_topology.go:137]: `func (e *evaluatorNetworkTopology) calculateParentHostUploadSuccessScore(dst, src string) float64 {`

File: scheduler/scheduling/evaluator/evaluator_network_topology.go [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+	// Host has not been scheduled, then it is scheduled first.
+	if uploadCount == 0 && uploadFailedCount == 0 {
+		return networkTopologyMaxScore
+	}
+
+	return float64(uploadCount-uploadFailedCount) / float64(uploadCount)
+}
+
+// networkTopologyCalculateFreeUploadScore 0.0~1.0 larger and better.
+func networkTopologyCalculateFreeUploadScore(host *resource.Host) float64 {
```
Comment (gaius-qi) [inline scheduler/scheduling/evaluator/evaluator_network_topology.go:153]: `func (e *evaluatorNetworkTopology) calculateFreeUploadScore(dst, src string) float64 {`

File: scheduler/scheduling/evaluator/evaluator_network_topology.go [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+	freeUploadCount := host.FreeUploadCount()
+	if ConcurrentUploadLimit > 0 && freeUploadCount > 0 {
+		return float64(freeUploadCount) / float64(ConcurrentUploadLimit)
+	}
+
+	return networkTopologyMinScore
+}
+
+// networkTopologyCalculateHostTypeScore 0.0~1.0 larger and better.
+func networkTopologyCalculateHostTypeScore(peer *resource.Peer) float64 {
```
Comment (gaius-qi) [inline scheduler/scheduling/evaluator/evaluator_network_topology.go:164]: `func (e *evaluatorNetworkTopology) calculateHostTypeScore(dst, src string) float64 {`

File: scheduler/scheduling/evaluator/evaluator_network_topology.go [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+		}
+
+		return networkTopologyMinScore
+	}
+
+	return networkTopologyMaxScore * 0.5
+}
+
+// calculateIDCAffinityScore 0.0~1.0 larger and better.
+func networkTopologyCalculateIDCAffinityScore(dst, src string) float64 {
```
Comment (gaius-qi) [inline scheduler/scheduling/evaluator/evaluator_network_topology.go:181]: `func (e *evaluatorNetworkTopology) calculateIDCAffinityScore(dst, src string) float64 {`

File: scheduler/scheduling/evaluator/evaluator_network_topology.go [hunk 14]
```diff
@@ -0,0 +1,233 @@
...
+
+	if strings.EqualFold(dst, src) {
+		return networkTopologyMaxScore
+	}
+
+	return networkTopologyMinScore
+}
+
+// networkTopologyCalculateMultiElementAffinityScore 0.0~1.0 larger and better.
+func networkTopologyCalculateMultiElementAffinityScore(dst, src string) float64 {
```
Comment (gaius-qi) [inline scheduler/scheduling/evaluator/evaluator_network_topology.go:194]: `func (e *evaluatorNetworkTopology) calculateMultiElementAffinityScore(dst, src string) float64 {`

File: scheduler/scheduling/evaluator/evaluator_network_topology.go [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+		}
+
+		score++
+	}
+
+	return float64(score) / float64(networkTopologyMaxElementLen)
+}
+
+// networkTopologyCalculateNetworkTopologyScore 0.0~1.0 larger and better.
+func (en *evaluatorNetworkTopology) networkTopologyCalculateNetworkTopologyScore(dst, src string) float64 {
```
Comment (gaius-qi) [inline scheduler/scheduling/evaluator/evaluator_network_topology.go:226]: `func (e *evaluatorNetworkTopology) calculateProbeScore(dst, src string) float64 {`

File: scheduler/scheduling/evaluator/evaluator_network_topology.go [hunk 16]
```diff
@@ -0,0 +1,233 @@
...
+		func(i, j int) bool {
+			return en.evaluate(parents[i], child, totalPieceCount) > en.evaluate(parents[j], child, totalPieceCount)
+		},
+	)
+
+	return parents
+}
+
+// The larger the value, the higher the priority.
+func (en *evaluatorNetworkTopology) evaluate(parent *resource.Peer, child *resource.Peer, totalPieceCount int32) float64 {
```
Comment (gaius-qi) [inline scheduler/scheduling/evaluator/evaluator_network_topology.go:105]: `func (e *evaluatorNetworkTopology) evaluate(parent *resource.Peer, child *resource.Peer, totalPieceCount int32) float64 {`

File: scheduler/scheduling/evaluator/evaluator_network_topology.go [hunk 17]
```diff
@@ -0,0 +1,233 @@
...
+	en := &evaluatorNetworkTopology{}
+
+	for _, opt := range options {
+		opt(en)
+	}
+	return en
+}
+
+// EvaluateParents sort parents by evaluating multiple feature scores.
+func (en *evaluatorNetworkTopology) EvaluateParents(parents []*resource.Peer, child *resource.Peer, totalPieceCount int32) []*resource.Peer {
```
Comment (gaius-qi) [inline scheduler/scheduling/evaluator/evaluator_network_topology.go:93]: `func (e *evaluatorNetworkTopology) EvaluateParents(parents []*resource.Peer, child *resource.Peer, totalPieceCount int32) []*resource.Peer {`

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: scheduler/scheduling/scheduling_test.go [hunk 18]
```diff
@@ -196,14 +197,19 @@ var (
 )
 
 func TestScheduling_New(t *testing.T) {
+	ctl := gomock.NewController(t)
+	defer ctl.Finish()
+	mockNetworkTopology := networktopologymocks.NewMockNetworkTopology(ctl)
 	tests := []struct {
 		name      string
 		pluginDir string
+		option    []evaluator.Option
```
Comment (gaius-qi) [inline scheduler/scheduling/scheduling_test.go:206]: options

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: scheduler/scheduling/evaluator/evaluator.go [hunk 19]
```diff
@@ -17,38 +17,109 @@
...
+	normalDistributionLen = 30
+
+	// When costs len is greater than or equal to 2,
+	// the last cost can be compared and calculated.
+	minAvailableCostLen = 2
+)
+
+type Evaluator struct{}
+
+type Evaluation interface {
```
Comment (gaius-qi) [inline scheduler/scheduling/evaluator/evaluator.go:65]: `Evaluator`

File: scheduler/scheduling/evaluator/evaluator.go [hunk 20]
```diff
@@ -17,38 +17,109 @@
...
+	// If the number of samples is greater than or equal to 30,
+	// it is close to the normal distribution.
+	normalDistributionLen = 30
+
+	// When costs len is greater than or equal to 2,
+	// the last cost can be compared and calculated.
+	minAvailableCostLen = 2
+)
+
+type Evaluator struct{}
```
Comment (gaius-qi) [inline scheduler/scheduling/evaluator/evaluator.go:63]: `evaluator`

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: scheduler/scheduling/evaluator/evaluator_base.go [hunk 21]
```diff
@@ -81,30 +58,30 @@ func (eb *evaluatorBase) EvaluateParents(parents []*resource.Peer, child *resour
...
+			return eb.evaluate(parents[i], child, totalPieceCount) > eb.evaluate(parents[j], child, totalPieceCount)
 		},
 	)
 
 	return parents
 }
 
 // The larger the value, the higher the priority.
-func evaluate(parent *resource.Peer, child *resource.Peer, totalPieceCount int32) float64 {
+func (eb *evaluatorBase) evaluate(parent *resource.Peer, child *resource.Peer, totalPieceCount int32) float64 {
```
Comment (gaius-qi) [inline scheduler/scheduling/evaluator/evaluator_base.go:69]: `func (eb *evaluatorBase)` => `func (e *evaluatorBase)`

**Review by gaius-qi — CHANGES_REQUESTED** [review]

**Review by gaius-qi — CHANGES_REQUESTED** [review]
LGTM

**Review by gaius-qi — APPROVED** [review]
LGTM

**Pull request merged**